### PR TITLE
Add support for .NET workload sets

### PR DESCRIPTION
**Pull request by joeloff**
# Description

Add support for generating workload set MSIs from a workload set NuGet package as described [here](https://github.com/dotnet/designs/pull/294).

Changes include

- Refactoring common properties for workload creation tasks into a base class since we now have two tasks
- Cleaning up the directory template to use fragments and share directory table authoring across multiple MSIs
- Additional tests to verify generated workload set installers
- Explicit test package versions in the unit test project. These are very specific versions and should rarely change. Placing them in versions.props doesn't provide much benefit.
- Additional package group type metadata to distinguish between workload manifest and workload set package groups for inserting to Visual Studio

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadPackMsi.wix.cs [hunk 1]
```diff
@@ -70,7 +70,7 @@ public override ITaskItem Build(string outputPath, ITaskItem[]? iceSuppressions
             candle.AddPreprocessorDefinition(PreprocessorDefinitionNames.InstallationRecordKey, $"InstalledPacks");
 
             if (!candle.Execute())
-            {
+            {
```
Comment (nagilson) [inline src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadPackMsi.wix.cs:73]:
```suggestion
            {
```

File: src/Microsoft.DotNet.Build.Tasks.Workloads.Tests/PackageTests.cs [hunk 2]
```diff
@@ -22,13 +18,35 @@ public void ItCanReadAManifestPackage()
         {
             string PackageRootDirectory = Path.Combine(BaseIntermediateOutputPath, "pkg");
 
-            TaskItem manifestPackageItem = new(Path.Combine(TestAssetsPath, "microsoft.net.workload.mono.toolchain.manifest-6.0.200.6.0.3.nupkg"));
+            TaskItem manifestPackageItem = new(Path.Combine(TestAssetsPath, "microsoft.net.workload.mono.toolchain.manifest-6.0.300.6.0.22.nupkg"));
```
Comment (nagilson) [inline src/Microsoft.DotNet.Build.Tasks.Workloads.Tests/PackageTests.cs:21]: Why was this changed to 300?

Comment (joeloff) [reply src/Microsoft.DotNet.Build.Tasks.Workloads.Tests/PackageTests.cs:21]: The files get locked during the MSI compilation, which means we either need to run all the tests sequentially from a single class. Using different packages removes the locking risk and avoid increasing the time it takes to run the tests. The full MSI tests can take close to 5 minutes per unit test.

File: src/Microsoft.DotNet.Build.Tasks.Workloads.Tests/Microsoft.DotNet.Build.Tasks.Workloads.Tests.csproj [hunk 3]
```diff
@@ -25,13 +25,15 @@
...
+    <PackageDownload Include="Microsoft.iOS.Templates" Version="[$(MicrosoftiOSTemplatesVersion)];[16.0.527]" />
     <PackageDownload Include="Microsoft.NET.Workload.Emscripten.net6.Manifest-8.0.100-preview.6" Version="[$(MicrosoftNETWorkloadEmscriptenManifest_80100Preview6Version)]"/>
     <!-- We can only test the task properly by building an actual workload end-to-end. EMSDK is the tiniest one available -->
     <PackageDownload Include="Microsoft.NET.Workload.Emscripten.Manifest-6.0.200" Version="[$(MicrosoftNETWorkloadEmscriptenManifest_60200Version)]" />
     <PackageDownload Include="Microsoft.NET.Runtime.Emscripten.2.0.23.Node.win-x64" Version="[$(MicrosoftNETRuntimeEmscripten2023Nodewin_x64)]" />
     <PackageDownload Include="Microsoft.NET.Runtime.Emscripten.2.0.23.Python.win-x64" Version="[$(MicrosoftNETRuntimeEmscripten2023Pythonwin_x64)]" />
     <PackageDownload Include="Microsoft.NET.Runtime.Emscripten.2.0.23.Sdk.win-x64" Version="[$(MicrosoftNETRuntimeEmscripten2023Sdkwin_x64)]" />
+    <PackageDownload Include="Microsoft.NET.Workloads.9.0.100" Version="[9.0.100-baseline.1.23464.1]" />
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.Build.Tasks.Workloads.Tests/Microsoft.DotNet.Build.Tasks.Workloads.Tests.csproj:36]: nit: We put all package versions into Versions.props / Directory.Packages.props

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/MsiTemplate/WorkloadSetProduct.wxs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    <!-- The new MSI is installed before the older version is removed. This is the fastest upgrade option. -->
+    <InstallExecuteSequence>
+      <!-- See https://docs.microsoft.com/en-us/windows/win32/msi/removeexistingproducts-action for other options -->
+      <RemoveExistingProducts After="InstallFinalize" />
+    </InstallExecuteSequence>
+
+    <Condition Message="A newer version of [ProductName] is alread installed.">
+      NOT WIX_DOWNGRADE_DETECTED
+    </Condition>
```
Comment (MiYanni) [inline src/Microsoft.DotNet.Build.Tasks.Workloads/src/MsiTemplate/WorkloadSetProduct.wxs:24]: nit:
```suggestion
    </Condition>
```

**Review by MiYanni — DISMISSED** [review]
I might not know the full context of how these changes operate, but the code looked really clean to me. Big thumbs up from me. 👍

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs [hunk 5]
```diff
@@ -208,329 +120,266 @@ public bool AllowMissingPacks
...
+                                new(msiOutputItem, BaseIntermediateOutputPath, BaseOutputPath, sdkFeatureBand, chip: null, machineArch: msiOutputItem.GetMetadata(Metadata.Platform), outOfSupport: IsOutOfSupportInVisualStudio) :
+                                new(msiOutputItem, BaseIntermediateOutputPath, BaseOutputPath, sdkFeatureBand, chip: msiOutputItem.GetMetadata(Metadata.Platform), outOfSupport: IsOutOfSupportInVisualStudio);
+                            string swixProj = swixProject.Create();
 
-                                    ITaskItem swixProjectItem = new TaskItem(swixProj);
-                                    swixProjectItem.SetMetadata(Metadata.SdkFeatureBand, $"{manifestPackage.SdkFeatureBand}");
-                                    swixProjectItem.SetMetadata(Metadata.PackageType, DefaultValues.PackageTypeMsiPack);
-                                    swixProjectItem.SetMetadata(Metadata.IsPreview, "false");
+                            ITaskItem swixProjectItem = new TaskItem(swixProj);
+                            swixProjectItem.SetMetadata(Metadata.SdkFeatureBand, $"{sdkFeatureBand}");
```
[the remark this answers is not in the document]
Comment (joeloff) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs:357]: yup, I'm embracing interpolation

[the remark this answers is not in the document]
Comment (MiYanni) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs:357]: The bold AND the brave only embrace interpolation so... fearlessly. 😜 I wonder if the IL ends up any different. I just hadn't seen someone use string interpolation for single-values before.

File: src/Microsoft.DotNet.Build.Tasks.Workloads.Tests/Microsoft.DotNet.Build.Tasks.Workloads.Tests.csproj [hunk 6]
```diff
@@ -43,12 +45,16 @@
...
     <Content Include="$(NuGetPackageRoot)microsoft.net.runtime.emscripten.2.0.23.node.win-x64\$(MicrosoftNETRuntimeEmscripten2023Nodewin_x64)\*.nupkg" Link="testassets\%(Filename)%(Extension)" CopyToOutputDirectory="PreserveNewest" Visible="false" />
     <Content Include="$(NuGetPackageRoot)microsoft.net.runtime.emscripten.2.0.23.python.win-x64\$(MicrosoftNETRuntimeEmscripten2023Pythonwin_x64)\*.nupkg" Link="testassets\%(Filename)%(Extension)" CopyToOutputDirectory="PreserveNewest" Visible="false" />
     <Content Include="$(NuGetPackageRoot)microsoft.net.runtime.emscripten.2.0.23.sdk.win-x64\$(MicrosoftNETRuntimeEmscripten2023Sdkwin_x64)\*.nupkg" Link="testassets\%(Filename)%(Extension)" CopyToOutputDirectory="PreserveNewest" Visible="false" />
+    <Content Include="$(NuGetPackageRoot)microsoft.net.workloads.9.0.100\9.0.100-baseline.1.23464.1\*.nupkg" Link="testassets\%(Filename)%(Extension)" CopyToOutputDirectory="PreserveNewest" Visible="false" />
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.Build.Tasks.Workloads.Tests/Microsoft.DotNet.Build.Tasks.Workloads.Tests.csproj:57]: Use the defined version properties for the four added content items.

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        protected override string BaseOutputName => Path.GetFileNameWithoutExtension(_package.PackagePath);
+
+        public WorkloadSetMsi(WorkloadSetPackage package, string platform, IBuildEngine buildEngine, string wixToolsetPath,
+            string baseIntermediatOutputPath) :
+            base(MsiMetadata.Create(package), buildEngine, wixToolsetPath, platform, baseIntermediatOutputPath)
+        {
+            _package = package;
+        }
+
+        public override ITaskItem Build(string outputPath, ITaskItem[]? iceSuppressions)
```
Comment (nagilson) [inline src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs:29]: What is an `iceSuppression`?

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            candle.AddSourceFiles(packageContentWxs,
+                EmbeddedTemplates.Extract("DependencyProvider.wxs", WixSourceDirectory),
+                EmbeddedTemplates.Extract("Directories.wxs", WixSourceDirectory),
+                EmbeddedTemplates.Extract("dotnethome_x64.wxs", WixSourceDirectory),
+                EmbeddedTemplates.Extract("WorkloadSetProduct.wxs", WixSourceDirectory));
+
+            // Extract the include file as it's not compilable, but imported by various source files.
+            EmbeddedTemplates.Extract("Variables.wxi", WixSourceDirectory);
+            
+            Guid upgradeCode = Utils.CreateUuid(UpgradeCodeNamespaceUuid, $"{_package.Identity};{Platform}");
```
Comment (nagilson) [inline src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs:58]: So, the `_package.Identity` will be the same for a given workload MSI, assuring the same upgrade code each time?  Who manages that identity?

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            string providerKeyName = $"Microsoft.NET.Workload.Set,{_package.SdkFeatureBand},{_package.PackageVersion},{Platform}";
+
+            // Set up additional preprocessor definitions.
+            candle.AddPreprocessorDefinition(PreprocessorDefinitionNames.UpgradeCode, $"{upgradeCode:B}");
+            candle.AddPreprocessorDefinition(PreprocessorDefinitionNames.DependencyProviderKeyName, $"{providerKeyName}");
+            candle.AddPreprocessorDefinition(PreprocessorDefinitionNames.SourceDir, $"{packageDataDirectory}");
+            candle.AddPreprocessorDefinition(PreprocessorDefinitionNames.SdkFeatureBandVersion, $"{_package.SdkFeatureBand}");
```
Comment (nagilson) [inline src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs:65]: Why doesn't the candle need the MSI version, does Light do that for us? If so, why is it also generated inside of `wix.targets`? Or that preprocessor definition just wouldnt be used, so it isn't added?

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs [hunk 10]
```diff
@@ -208,329 +120,266 @@ public bool AllowMissingPacks
             set;
         } = false;
 
-        public override bool Execute()
+        protected override bool ExecuteCore()
         {
-            try
+            // TODO: trim out duplicate manifests.
```
Comment (nagilson) [inline src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs:125]: Should this TODO have been completed?

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs [hunk 11]
```diff
@@ -208,329 +120,266 @@ public bool AllowMissingPacks
...
+            List<WorkloadManifestMsi> manifestMsisToBuild = new();
+            HashSet<SwixComponent> swixComponents = new();
+            HashSet<SwixPackageGroup> swixPackageGroups = new();
+            Dictionary<string, BuildData> buildData = new();
+            Dictionary<string, WorkloadPackGroupPackage> packGroupPackages = new();
+
+            // First construct sets of everything that needs to be built. This includes
+            // all the packages (manifests and workload packs) that need to be extracted along
+            // with the different installer types. 
+            foreach (ITaskItem workloadManifestPackageFile in WorkloadManifestPackageFiles)
```
Comment (nagilson) [inline src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs:136]: How does this `WorkloadManifestPackageFiles` get instantiated?

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs [hunk 12]
```diff
@@ -72,15 +46,6 @@ public bool EnableSideBySideManifests
             set;
         } = false;
```
Comment (nagilson) [inline src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs:47]: Should this default to `true` now? If not, why? Because we still need to build older SDK workloads?

**Review by nagilson — APPROVED** [review]
I have more questions than really anything else... I don't understand this all too well, but it seems reasonable given the knowledge I do have. If there is someone out there with this domain knowledge, it'd be best to have their review too.

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs (hunk 7 above)
Comment (MiYanni) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs:29]: I said the same thing to myself but then I saw what they were in the PR itself here: https://github.com/dotnet/arcade/pull/14083/files#diff-00d1756db7ead438b087502a3086ab2adc018742e30374e23f7f70e32f1aaabfR49-R56

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs (hunk 8 above)
Comment (joeloff) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs:58]: That falls to workload owners, but in identity in the NuGet API includes the package ID and package version. Previously when we supported upgradable manifests, we only looked at the platform and package ID.

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs (hunk 7 above)
Comment (joeloff) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs:29]: Think of an ICE as a static analysis rule for MSIs. They are intended to validate that you don't break certain rules that can lead to issues when the MSI is actually being executed. Most of them should be adhered to, but there are some you can/want to suppress.

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs (hunk 10 above)
Comment (joeloff) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs:125]: Not yet - that's been there for a few iterations. I need to review the code again. Possibly the answer will eventually be we can remove this. It comes from when we were building multiple manifests and at one point ended up with the same manifest appearing more than once. and there was an idea that the task could be hardened to remove duplicate items.

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs (hunk 11 above)
Comment (joeloff) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs:136]: One a repo like runtime produces the normal workload manifest packages, those files are put into an item group and passed as a parameter, e.g. https://github.com/dotnet/runtime/blob/e8c8ab888c125c7ccdf48d36137f6a7d243bf53a/src/workloads/workloads.csproj#L144 and  https://github.com/dotnet/runtime/blob/e8c8ab888c125c7ccdf48d36137f6a7d243bf53a/src/workloads/workloads.csproj#L156

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs (hunk 12 above)
Comment (joeloff) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkload.wix.cs:47]: yup, just avoiding messing with the default behavior. Eventually we may want other repos to switch over, but changes to the defaults impact all the consumers f the task

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs (hunk 9 above)
Comment (joeloff) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs:65]: Candle is the compiler, so the MSI version comes as a preprocessor variable when the MSI source file is compiled, hence why it's passed to the compiler on the commandline. Light is the linker, so it processes the generated object files which has the MSI version and therefore doesn't need it again.

Comment (nagilson) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/Msi/WorkloadSetMsi.wix.cs:65]: That makes sense, thank you!

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkloadSet.wix.cs [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        public string WorkloadSetMsiVersion
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// Set of NuGet packages containing workload sets.
+        /// </summary>
+        public ITaskItem[] WorkloadSetPackageFiles
```
Comment (dsplaisted) [inline src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkloadSet.wix.cs:32]: Do we ever expect there to be more than one of these packages passed to the task?

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkloadSet.wix.cs [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+                    swixProjectItem.SetMetadata(Metadata.PackageType, DefaultValues.PackageTypeMsiWorkloadSet);
+                    swixProjectItem.SetMetadata(Metadata.IsPreview, "false");
+
+                    lock (swixProjectItems)
+                    {
+                        swixProjectItems.Add(swixProjectItem);
+                    }
+                });
+
+                _ = Parallel.ForEach(swixPackageGroups, swixPackageGroup =>
```
Comment (dsplaisted) [inline src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkloadSet.wix.cs:98]: Is there any point in doing this in parallel, since all the work inside it is serialized in a `lock` block?

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkloadSet.wix.cs [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+                    lock (swixProjectItems)
+                    {
+                        swixProjectItems.Add(PackageGroupSwixProject.CreateProjectItem(swixPackageGroup, BaseIntermediateOutputPath, BaseOutputPath,
+                            DefaultValues.PackageTypeWorkloadSetPackageGroup));
+                    }
+                });
+            }
+
+            Msis = msiItems.ToArray();
+            SwixProjects = swixProjectItems.ToArray();
```
Comment (dsplaisted) [inline src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkloadSet.wix.cs:109]: I think for the baseline workload sets, which will be created in the installer repo, we will want to create the MSIs, but we won't insert the authoring into Visual Studio.  Will we be able to filter out the VS authoring appropriately?

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/MsiTemplate/WorkloadSetProduct.wxs [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    <Upgrade Id="$(var.UpgradeCode)">
+      <UpgradeVersion Maximum="$(var.ProductVersion)" MigrateFeatures="yes" IncludeMinimum="no" Property="WIX_UPGRADE_DETECTED" />
+      <UpgradeVersion Minimum="$(var.ProductVersion)" IncludeMinimum="no" OnlyDetect="yes" Property="WIX_DOWNGRADE_DETECTED" />
+    </Upgrade>
+
+    <!-- The new MSI is installed before the older version is removed. This is the fastest upgrade option. -->
+    <InstallExecuteSequence>
+      <!-- See https://docs.microsoft.com/en-us/windows/win32/msi/removeexistingproducts-action for other options -->
+      <RemoveExistingProducts After="InstallFinalize" />
+    </InstallExecuteSequence>
```
Comment (dsplaisted) [inline src/Microsoft.DotNet.Build.Tasks.Workloads/src/MsiTemplate/WorkloadSetProduct.wxs:20]: Is the upgrade logic needed for workload sets?  They are side-by-side so I don't think they should ever be upgraded.

**Review by dsplaisted — APPROVED** [review]

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/MsiTemplate/WorkloadSetProduct.wxs (hunk 16 above)
Comment (joeloff) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/MsiTemplate/WorkloadSetProduct.wxs:20]: We can remove that.

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkloadSet.wix.cs (hunk 15 above)
Comment (joeloff) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkloadSet.wix.cs:109]: Yes. Right now the tasks generate the projects for those, but it's up to the repo to decided whether or not it wants to build those projects. Here's an example where the runtime builds the swix projects: https://github.com/dotnet/runtime/blob/ccc475ed5646abe71ef596702d06e78e34981736/src/workloads/workloads.csproj#L179

Alternatively we could add a parameter that controls generating the swixprojects.

File: src/Microsoft.DotNet.Build.Tasks.Workloads.Tests/SwixPackageTests.cs [hunk 17]
```diff
@@ -34,41 +35,24 @@ public void ItThrowsIfPackageRelativePathExceedsLimit()
...
-            ITaskItem item = msi.Build(MsiOutputPath);
-
-            Assert.Equal("Microsoft.iOS.Templates.15.2.302-preview.14.122", item.GetMetadata(Metadata.SwixPackageId));
-        }
-
-        [WindowsOnlyFact]
-        public void ItOnlyIncludesDefinedPropertiesForMsiPackages()
+        [WindowsOnlyTheory]
+        [InlineData("16.0.527")]
+        public void ItOnlyIncludesDefinedPropertiesForMsiPackages(string packageVersion)
```
[the remark this answers is not in the document]
Comment (joeloff) [reply src/Microsoft.DotNet.Build.Tasks.Workloads.Tests/SwixPackageTests.cs:40]: No, I had multiple tests before I created the PR, but started cleaning up. Good catch

File: src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkloadSet.wix.cs (hunk 13 above)
Comment (joeloff) [reply src/Microsoft.DotNet.Build.Tasks.Workloads/src/CreateVisualStudioWorkloadSet.wix.cs:32]: Probably not, but if we ever needed it, we won't need an Arcade update.

**Comment by joeloff** [comment]
Any other feedback or are we good to merge?

**Comment by dsplaisted** [comment]
Good to merge as far as I'm concerned

**Review by dsplaisted — APPROVED** [review]
